packages/flutter_test/lib/src/matchers.dart

```diff
@@ -1973,10 +2237,40 @@ class _MatchesSemanticsData extends Matcher {
       description.add(' with tooltip: $tooltip');
     }
     if (actions != null) {
-      description.add(' with actions: ').addDescriptionOf(actions);
+      final List<SemanticsAction> expectedActions = <SemanticsAction>[];
+      final List<SemanticsAction> notExpectedActions = <SemanticsAction>[];
+      for (final MapEntry<ui.SemanticsAction, bool> actionEntry in actions.entries) {
+        if (actionEntry.value) {
+          expectedActions.add(actionEntry.key);
+        } else {
+          notExpectedActions.add(actionEntry.key);
+        }
+      }
+
+      if (expectedActions.isNotEmpty) {
+        description.add(' with actions: ').addDescriptionOf(expectedActions);
+      }
+      if (notExpectedActions.isNotEmpty) {
+        description.add(' without actions: ').addDescriptionOf(notExpectedActions);
+      }
     }
     if (flags != null) {
-      description.add(' with flags: ').addDescriptionOf(flags);
+      final List<SemanticsFlag> expectedFlags = <SemanticsFlag>[];
+      final List<SemanticsFlag> notExpectedFlags = <SemanticsFlag>[];
+      for (final MapEntry<SemanticsFlag, bool> flagEntry in flags.entries) {
+        if (flagEntry.value) {
+          expectedFlags.add(flagEntry.key);
+        } else {
+          notExpectedFlags.add(flagEntry.key);
+        }
+      }
```

Comment:
This could be:

```dart
final expectedFlags = flags.entries.where((e) => e.value).map((e) => e.key);
final notExpectedFlags = flags.entries.where((e) => !e.value).map((e) => e.key);
```

I'll let you decide whether this is more readable, but I generally prefer higher order functions for situations like this.

Reply:
This goes for the `actions` section above as well

Reply:
Good call! Ended up not _quite_ as pretty as your suggestions due to the `always_specify_types` lint, but I still think it came out more readable with the higher order functions route.